### DISCUSSION TITLE
L4CasADi: PyTorch Models in CasADi

### DISCUSSION THREAD

**Discussion by Tim-Salzmann** [casadi/casadi#3275] [Show and tell]
Hi CasADi Community,

If you are looking for a way to integrate data-driven PyTorch Models in a CasADi computational/symbolic graph, [L4CasADi](https://github.com/Tim-Salzmann/l4casadi) will efficiently do just that.

Key Features are

- All jit traceable PyTorch models are supported.
- GPU / CUDA support (MPS for Mac M planned with next PyTorch update).
- C Code generation with an L4CasADi component in the CasADi graph.
- CasADi solvers work with L4CasADi components: [Example](https://github.com/Tim-Salzmann/l4casadi/blob/main/examples/simple_nlp.py)
- Acados can use CasADi graphs with L4CasADi components: [Example](https://github.com/Tim-Salzmann/l4casadi/blob/main/examples/acados.py)

L4CasADi can be easily integrated with CasADi:
```
 l4c_model = l4c.L4CasADi(pyTorch_model, has_batch=True, device='cpu')  # device='cuda' for GPU 
  
 x_sym = cs.MX.sym('x', 2, 1) 
 y_sym = l4c_model(x_sym) 
 f = cs.Function('y', [x_sym], [y_sym]) 
 df = cs.Function('dy', [x_sym], [cs.jacobian(y_sym, x_sym)]) 
 ddf = cs.Function('ddy', [x_sym], [cs.hessian(y_sym, x_sym)[0]]) 
  
 x = cs.DM([[0.], [2.]]) 
 print(l4c_model(x)) 
 print(f(x)) 
 print(df(x)) 
```

More information can be found [here](https://github.com/Tim-Salzmann/l4casadi).

Any Feedback is appreciated!

**Comment by nacereddinemoumai** [comment]
@Tim-Salzmann Thank you for the explanations. However, I have a question. I have a trained PyTorch model that takes a tensor of shape [batch, window, nu+nx] as input to predict the next state of my system. I am encountering a problem when constructing an L4CasADi model from this PyTorch model. The construction works, but when I pass a tensor as input, I get an error: **ValueError: For batched PyTorch models only vector inputs are allowed**. My question is: how can I resolve this issue?
`
# Construct L4CasADi Model from PyTorch Model
from casadi import *
import l4casadi as l4c
l4c_model = l4c.L4CasADi(model, model_expects_batch_dim=True, device='cpu')

tx = MX.sym('Text',5)
beta = MX.sym('beta',5)
Tz = MX.sym('tin',5)
input = horzcat(tx,beta,Tz)
print(input.shape)
y_sym = l4c_model(input) 
`
![Additionally, here is the architecture of the model:](https://github.com/user-attachments/assets/0a2b4db6-3492-4be3-9b27-4ff7c293768c)

**Reply by Tim-Salzmann** [reply]
Hi, please ask your questions here: https://github.com/Tim-Salzmann/l4casadi/issues

You can solve this by reshaping your CasADi input to a vector and, inside your PyTorch model, reshaping it back to the expected dimensions. If this is not sufficient, please open an issue under the above link.

**Reply by nacereddinemoumai** [reply]
@Tim-Salzmann  The solution you proposed works. Thank you, sir!
